client/go/outline/iptable/iptable_listener_test.go

```diff
@@ -0,0 +1,299 @@
+
+	mockConn1.readChan <- packetFrom1
+	mockConn2.readChan <- packetFrom2
+	mockConnDefault.readChan <- packetFromDefault
+
+	expectedResponses := map[string][]byte{
+		mockConn1.addr.String():       packetFrom1,
+		mockConn2.addr.String():       packetFrom2,
+		mockConnDefault.addr.String(): packetFromDefault,
+	}
+	receivedResponses := make(map[string][]byte)
+
+	readBuf := make([]byte, 1024)
+	for i := 0; i < len(expectedResponses); i++ {
+		n, addr, readErr := conn.ReadFrom(readBuf)
+		// Check for context timeout or other fatal errors
+		if ctx.Err() != nil {
+			t.Fatalf("Context timed out while waiting for ReadFrom: %v", ctx.Err())
+		}
+		require.NoError(t, readErr, "ReadFrom failed on iteration %d", i)
+
+		responseBytes := make([]byte, n)
+		copy(responseBytes, readBuf[:n])
+		receivedResponses[addr.String()] = responseBytes
+	}
+
+	assert.Equal(t, expectedResponses, receivedResponses, "Mismatch in received responses")
+
+	// Test SetDefault to nil and try writing (should fail if IP not in table)
+	iptableListener.SetDefault(nil)
+	conn2, err := iptableListener.ListenPacket(ctx) // iptableListener now has nil default
+	require.NoError(t, err, "ListenPacket (conn2) failed")
+	defer conn2.Close()
+
+	_, err = conn2.WriteTo(testData, destDefault) // Should use the new nil default listener
+	require.Error(t, err, "conn2.WriteTo to destDefault should fail as default listener is nil")
+	t.Logf("Error from WriteTo with nil default: %v", err)
+
+	// Test writing to an address that IS in the table, should still work
+	_, err = conn2.WriteTo(testData, dest1)
+	require.NoError(t, err, "conn2.WriteTo to dest1 should still work")
+	select {
+	case received := <-mockConn1.writeChan:
+		assert.Equal(t, testData, received, "mockConn1 (via conn2) received unexpected data")
+	case <-time.After(100 * time.Millisecond):
+		t.Fatal("mockConn1 (via conn2) did not receive data after SetDefault(nil)")
+	}
+}
+
+func TestIPTablePacketListener_IPv4IPv6Routing(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	table := NewIPTable[transport.PacketListener]()
+
+	mockConnV4Specific := newMockPacketConn(t, "192.0.2.1", "127.0.0.1")
+	mockConnV6Specific := newMockPacketConn(t, "2001:db8:1::1", "::1")
+	mockConnDefault := newMockPacketConn(t, "10.0.0.1", "127.0.0.1")
+
+	require.NoError(t, table.AddPrefix(netip.MustParsePrefix("192.0.2.0/24"), &mockPacketListener{conn: mockConnV4Specific}))
+	require.NoError(t, table.AddPrefix(netip.MustParsePrefix("2001:db8:1::/48"), &mockPacketListener{conn: mockConnV6Specific}))
+
+	defaultListener := &mockPacketListener{conn: mockConnDefault}
+	iptableListener, err := NewIPTablePacketListener(table, defaultListener)
+	require.NoError(t, err, "NewIPTablePacketListener failed")
+
+	conn, err := iptableListener.ListenPacket(ctx)
+	require.NoError(t, err, "ListenPacket failed")
+	defer conn.Close()
+
+	testData := []byte("ipv4-ipv6-test")
+
+	// 1. IPv4
```

Comment:
This can be separated using sub-tests, for example: 
```
t.Run("IPv4", func() {
  destV4PureAddrStr := "..."
  ...
})

t.Run("IPv6", func() {
  ...
})
```